theano/gpuarray/extra_ops.py

```diff
@@ -451,24 +451,24 @@ def c_support_code_struct(self, node, nodename):
         return super(GpuCumsum, self).c_support_code_struct(node, nodename) + code
 
 
+@register_opt('fast_compile')
 @op_lifter([CumsumOp])
-def use_gpu_cumsumop(node, ctx_name):
-    if node.inputs[0].dtype == 'float32':
-        axis = node.op.axis
-        x = node.inputs[0]
+@register_opt2([CumsumOp], 'fast_compile')
+def use_gpu_cumsumop(op, ctx_name, inputs, outputs):
+    if inputs[0].dtype == 'float32':
+        axis = op.axis
+        x = inputs[0]
 
         if axis is not None and x.ndim > GpuCumsum.SUPPORTED_NDIMS:
             return None
 
         if axis is None and x.ndim > 1:
             x = x.flatten()
 
-        x = GpuFromHost(ctx_name)(x)
+        x = as_gpuarray_variable(x, ctx_name)
```

Comment:
`.reshape()` gives a CPU Op because we usually build the graph on the CPU, take gradients on the CPU, etc., and only transfer to the GPU during optimizations. That way, only the CPU Ops need to define gradients, for instance.
Just try the following:
```python
>>> x = theano.shared(np.ones((2, 3)))
>>> print(x.type)
GpuArrayType<None>(float64, (False, False))
>>> y = x.reshape((3, 2))
>>> print(y.type)
TensorType(float64, matrix)
```
This is something that could change in the future, but for the moment it is the way it is implemented.
To avoid that issue, you can use `GpuReshape` directly, which is what I suggested earlier.